### PR TITLE
Refactors based on jrddunbr's ideas and general cleanup

### PR DESCRIPTION
**Pull request by empireu**
- Makes Scale generic and moves units to more appropriate packages
- Moves Thermal Constructs and simulator to a separate file

### CODE REVIEW

File: src/main/kotlin/org/ageseries/libage/sim/measurement/Scale.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    fun map(t: Double): Double = factor * t + base
+
+    /** Given a temperature [t] in this unit, return its value in standard units. */
+    fun unmap(t: Double): Double = (t - base) / factor
+
+    /**
+     * Give a human-readable string of a value in this scale.
+     */
+    fun display(t: Double): String = "${t}${displayUnit}"
+}
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/measurement/Scale.kt:19]: Missing newline at end of file.

File: src/main/kotlin/org/ageseries/libage/sim/measurement/ThermalUnits.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    val CELSIUS = CENTIGRADE
+    /**
+     * Milligrade; zero as Grade, but with steps one-thousandth of a grade.
+     */
+    val MILLIGRADE = Scale(10.0, -2731.5, "mG")
+    /**
+     * Fahrenheit; 32 is zero grade, 212 is one grade.
+     */
+    val FAHRENHEIT = Scale(9.0/5.0, -491.67, "°F")
+}
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/measurement/ThermalUnits.kt:46]: Missing newline at end of file.

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+         */
+        fun from(temp: Double, scale: Scale) = Temperature(scale.unmap(temp))
+    }
+}
+
+/**
+ * "Standard" temperature, in Kelvin.
+ *
+ * This is the "ST" of STP, as reported in laboratory conditions.
+ *
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt:49]: Still doesn't cover _what_ STP it is. There are multiple standards. Please specify the centigrade equivalent in the comment.

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Simulation.kt [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            val overallCond = sqrt(cond * body.surfaceArea * body.mass.material.thermalConductivity)
+            val power = overallCond * deltaT
+            deltaE[body] = deltaE.getOrDefault(body, 0.0) + power * dt
+        }
+
+        deltaE.entries.forEach { (body, delta) ->
+            body.mass.energy += delta
+        }
+    }
+}
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Simulation.kt:134]: Missing newline.

File: src/main/kotlin/org/ageseries/libage/sim/constant/Material.kt [hunk 5]
```diff
@@ -83,7 +85,7 @@ Air:        0.024
 Source: http://hyperphysics.phy-astr.gsu.edu/hbase/Tables/thrcn.html
 */
 
-    abstract val specficHeat: Double
+    abstract val specificHeat: Double
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/constant/Material.kt:88]: Nice catch.

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    val celsius get() = to(ThermalUnits.CELSIUS)
+    val milligrade get() = to(ThermalUnits.MILLIGRADE)
+    val fahrenheit get() = to(ThermalUnits.FAHRENHEIT)
+
+    operator fun plus(rhs: Temperature) = Temperature(kelvin + rhs.kelvin)
+    operator fun minus(rhs: Temperature) = Temperature(kelvin - rhs.kelvin)
+
+    // Would it make sense to multiply 2 temps?
+    operator fun times(rhs: Double) = Temperature(kelvin * rhs)
+    operator fun div(rhs: Double) = Temperature(kelvin / rhs)
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt:33]: No, it doesn't really make sense in most cases imo.

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        // W/K
+        val distCondA = a.material.thermalConductivity * contactPoint * distance
+        val distCondB = b.material.thermalConductivity * (1.0 - contactPoint) * distance
+        val overallCond = (distCondA * distCondB * conductance).pow(1.0 / 3.0)
+        // W
+        val power = deltaT * overallCond
+        // J
+        return power * dt
+    }
+}
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt:124]: Missing newline

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+import org.ageseries.libage.sim.measurement.ThermalUnits
+import kotlin.math.pow
+
+/**
+ * A temperature.
+ *
+ * The inner unit type is always Kelvin; conversions are available as functions and properties.
+ */
+@JvmInline
+value class Temperature(val kelvin: Double) {
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt:14]: Might want `open` if we extend it as a serializable.

File: src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    var temperature: Temperature
+        get() = Temperature(energy / mass / material.specificHeat)
+        set(value) {
+            energy = value.kelvin * mass * material.specificHeat
+        }
+
+    override fun toString() = "<Thermal Mass $material ${mass}kg ${energy}J $temperature>"
+}
+
+data class ThermalConnectionParameters(
```
Comment (jrddunbr) [inline src/main/kotlin/org/ageseries/libage/sim/thermal/Constructs.kt:82]: Might want to use `open` here if we want to extend it with a serializable version.

**Review by jrddunbr — CHANGES_REQUESTED** [review]
Some nits.

**Comment by jrddunbr** [comment]
Merge conflicts.

**Comment by Grissess** [comment]
This _may_ have been accidentally rendered moot by my own pushes on #18 , but I'm open to cherry-picking some of these in anyway.

And some other changes as well: clarifying STP is on me, where we're using the IUPAC standard at 0C (NIST prefers 20C). Realistically, that value in Kelvin is exactly the difference between the zero points on centigrade and Kelvin :)

**Pull request closed without merging**